MovingBlocks/DestinationSol
# Added intitial support for animated sprites.

# Description
This pull request implements basic sprite-sheet based animation as a ```Drawable``` inheriting from ```RectSprite```. Sprite management has been moved into the ```SpriteManager``` class and all explicit references the the ```RectSprite``` class constructor have been updated to use the various ```SpriteManager.createSprite``` overloads instead.

# Testing
## Checking for bugs
- Run the game
- Start a new game - any settings will do
- Check that nothing looks different/broken (are the scales wrong? do objects even appear? etc.)
- Check that the collision shapes still align to their appearences (try bumping into things)
## Testings that sprite animations work
- Getting some animations:
I have created some sample animations to test that this feature works, they can be found on the "[animated-sprites-example](https://github.com/BenjaminAmos/DestinationSol/tree/animated-sprites-example)" branch on my fork. They are modified versions of the existing sprites. I may add a couple more of these to test various sizes and formats.
    - Download the [drome.png](https://github.com/BenjaminAmos/DestinationSol/blob/animated-sprites-example/modules/core/assets/ships/drome/drome.png) and [dromeAnimation.json](https://github.com/BenjaminAmos/DestinationSol/blob/animated-sprites-example/modules/core/assets/ships/drome/dromeAnimation.json) files and place them in the ```modules/core/assets/ships/drome``` directory.
    - Download the [station.png](https://github.com/BenjaminAmos/DestinationSol/blob/animated-sprites-example/modules/core/assets/ships/station/station.png) and [stationAnimation.json](https://github.com/BenjaminAmos/DestinationSol/blob/animated-sprites-example/modules/core/assets/ships/station/stationAnimation.json) files and place them in the ```modules/core/assets/ships/station``` directory.
    - Download the [rockyGround.json](https://github.com/BenjaminAmos/DestinationSol/blob/animated-sprites-example/modules/core/assets/grounds/rocky/rockyGround.json), [rockyGround_ff_2.png](https://github.com/BenjaminAmos/DestinationSol/blob/animated-sprites-example/modules/core/assets/grounds/rocky/rockyGround_ff_2.png) and [rockyGround_ff_2Animation.json](https://github.com/BenjaminAmos/DestinationSol/blob/animated-sprites-example/modules/core/assets/grounds/rocky/rockyGround_ff_2Animation.json) files and place them in the ```modules/core/assets/grounds/rocky``` directory.
    - Download the [bulletProjectile.png](https://github.com/BenjaminAmos/DestinationSol/blob/animated-sprites-example/modules/core/assets/textures/projectiles/bulletProjectile.png) and [bulletProjectileAnimation.json](https://github.com/BenjaminAmos/DestinationSol/blob/animated-sprites-example/modules/core/assets/textures/projectiles/bulletProjectileAnimation.json) files and place them in the ```modules/core/assets/textures/projectiles``` directory.
- Run the game
- Start a new game - any settings will do (unless stated otherwise)
- Testing animations:
    - Drome animations:
        - Fly down to any planet and find a drome (the planetary merchant station)
        - Check if the drome has an animated arrow moving up and down above it
        - Check if the drome has each of the letters of the sign lit in pairs (except the last "$" letter) over a short duration (the animation is deliberatly sped-up)
    - Station animations:
        - When you start a new game, you should spawn at a station
        - Check if the station has some pulsating green 'lights' in front of it (not actual lights)
    - Rocky ground tile animation:
        - Fly down to any rocky planet and find a piece of terrain that has a metal box inside it (look at the texture for a reference)
        - Check that there are moving waves coming out of an 'antenna' that extrudes partially out of the tile
        - Check that there are rapidly changing symbols shown on the face of the underground box
    - Bullet projectile animation:
        - Start a new game with the "Loaded Imperial Large" ship
        - Press "left-shift" (or whatever key was configured) to activate the slow-down time in-game ability
        - Click the left or right mouse buttons (or any configured equivalent) to shoot a projectile
        - Check that the projectile changes colour rapidly once fired

# Notes
This should not break existing sprites, although they have been upgraded to use the SpriteManager class so that they can use animated sprites without any other changes being needed. All the sprites produced by ```SpriteManager.createSprite``` are ```AnimatedRectSprites``` although this can be changed. You can use the ```SpriteManager.createStaticSprite``` method to create ```RectSprite``` objects directly.

~This pull request also adds a file directly from the LibGDX source tree, as it was extremely useful as a utility class for inplmenting these animations and the current LibGDX version referenced by Destination Sol was too old (it only supported ```TextureRegion``` objects). If absolutely essential, I can write a similar class (although fixed to only use ```AtlasRegion``` objects as frames) to replace this. This should not be needed if issue #312 is implemented, as the generic form of the class required should most certainly be in any version that is v1.9.5 or above.~ The Animation class from LibGDX is now used directly.

<!-- ### Pre Pull Request Checklist:
When scanning the code with SonarLint, just don't introduce any new issues, and maybe even fix a few existing ones
- [ ] Code has been scanned with [SonarLint](http://www.sonarlint.org/intellij/)
- [ ] Methods have been annotated with @ Nullable and @ Nonnull ([Read more](https://github.com/MovingBlocks/DestinationSol/wiki/@Nullable-&-@Nonnull-Quickstart))
- [ ] There are no errors present in the project
- [ ] Code has been formatted and indented
- [ ] Methods have appropriate Javadoc ([How to write Javadoc](http://www.oracle.com/technetwork/java/javase/documentation/index-137868.html))
